### ISSUE TITLE
Binding issues with ContextMenu and CompositeCollection

### ISSUE THREAD

**Issue by michaelmairegger** [MaterialDesignInXAML/MaterialDesignInXamlToolkit#3464]
### Bug explanation

I have XAML Binding Failures with MenusAndToolBars.xaml if I change the lines https://github.com/MaterialDesignInXAML/MaterialDesignInXamlToolkit/blob/master/MainDemo.Wpf/MenusAndToolBars.xaml#L355-L360 with the follwogin content:

    <ContextMenu>
      <ContextMenu.ItemsSource>
        <CompositeCollection>
          <MenuItem Header="Hello World" />
          <MenuItem Header="Clickety Click">
            <MenuItem Header="Clackety Clack" />
          </MenuItem>
        </CompositeCollection>
      </ContextMenu.ItemsSource>
    </ContextMenu>

saying:

> Severity	Count	Data Context	Binding Path	Target	Target Type	Description	File	Line	Project
Error	2	null	(0)	MenuItem.Height	Double	Cannot find source: RelativeSource FindAncestor, AncestorType='System.Windows.Controls.Primitives.MenuBase', AncestorLevel='1'.			

Do you have an idea how to fix this issue?

### Version

4.9.0

**Comment by nicolaihenriksen** [comment]
@michaelmairegger Is there a particular reason you choose the `ContextMenu.ItemsSource` property instead of just the XAML below?

The `ItemsSource` property is often used when the objects you have are not UI elements, and a `DataTemplate` is the used to visualize these "data" items. In your case, you are putting actual UI elements into the collection. I suspect this is also the reason for the binding error. Without having verified my theory, I would assume that the `MenuItem` instances you put in the `CompositeCollection` are not (at least not initially) part of the visual tree, but only the logical tree, and the binding which attempts to travel up the visual tree to find the root `ContextMenu` (i.e. `MenuBase`) then ends up failing.

```xaml
<ContextMenu>
  <MenuItem Header="Hello World" />
  <MenuItem Header="Clickety Click">
    <MenuItem Header="Clackety Clack" />
  </MenuItem>
</ContextMenu>
```

or

```xaml
<ContextMenu>
  <ContextMenu.Items>
      <MenuItem Header="Hello World" />
      <MenuItem Header="Clickety Click">
        <MenuItem Header="Clackety Clack" />
      </MenuItem>
  </ContextMenu.Items>
</ContextMenu>
```

**Comment by michaelmairegger** [comment]
@nicolaihenriksen Yes there is a reason. I have a "BaseMenu" that contains multiple MenuItems that are reused within many views, e.g.

    <UserControl.Resources>
      <CompositeCollection x:Key="BaseMenu">
        <MenuItem Header="Add" />
        <MenuItem Header="Edit" />
        <MenuItem Header="Delete" />
      </CompositeCollection>
    </UserControl.Resources>

    <ContextMenu>
      <ContextMenu.ItemsSource>
        <CompositeCollection >
          <CollectionContainer Collection="{StaticResource BaseMenu}" />
          <Separator></Separator>
          <MenuItem Header="Hello World" />
          <MenuItem Header="Clickety Click">
            <MenuItem Header="Clackety Clack" />
          </MenuItem>
        </CompositeCollection>
      </ContextMenu.ItemsSource>
    </ContextMenu>

**Comment by binarycow** [comment]
@michaelmairegger - try this:

```xaml
<ContextMenu>
  <ContextMenu.ItemsSource>
    <CompositeCollection >
      <CollectionContainer Collection="{Binding Source={StaticResource BaseMenu}}" /> <-- Note the binding
      <Separator />
      <MenuItem Header="Hello World" />
      <MenuItem Header="Clickety Click">
        <MenuItem Header="Clackety Clack" />
      </MenuItem>
    </CompositeCollection>
  </ContextMenu.ItemsSource>
</ContextMenu>
```

See here: https://stackoverflow.com/a/40147645

**Comment by michaelmairegger** [comment]
Unfortunately it is not working. I created a workaround by modifying the style and removing the two lines that created the issue

**Comment by MichelMichels** [comment]
> Unfortunately it is not working. I created a workaround by modifying the style and removing the two lines that created the issue

Could you tell us which lines?

**Comment by MichelMichels** [comment]
Thank you very much for your reply. I'll try to investigate this further.

**Comment by michaelmairegger** [comment]
Ok thanks for the information. I will try it on my end, and add a branch to my fork that replicates the issue.

**Comment by michaelmairegger** [comment]
I have added your code to MenuAndToolBars.xaml and the following gif shows the behavior

![Issue 3464](https://github.com/MaterialDesignInXAML/MaterialDesignInXamlToolkit/assets/4341039/68436a51-f305-44a6-bd57-e55ab93ab3b0)

**Comment by MichelMichels** [comment]
Apologies, my mistake, I was expecting a more blocking runtime exception. Now, I see the same problem you are seeing. I'll look into what the effect is of removing those lines you mentioned.

**Comment by MichelMichels** [comment]
This only seems to be an issue when used with a `ContextMenu`, a regular `Menu` does not give this error. I'm not sure why at this moment.

**Comment by michaelmairegger** [comment]
Thanks. I think the main issue will be that CompositeCollection is not part of the visual tree

**Comment by MichelMichels** [comment]
Yes, this indeed seems to be the case. The `AdornerDecorator` is not the problem. I tested this.

I don't know if we will be able to resolve this as the default style for `MenuItem` is expecting a `MenuBase` usercontrol as its parent somewhere in the tree. I'm trying to get myself familiar with `CompositeCollection`, but its the first time that I work with this class.

**Comment by michaelmairegger** [comment]
> 
> ```
> <UserControl.Resources>
>   <CompositeCollection x:Key="BaseMenu">
>     <MenuItem Header="Add" />
>     <MenuItem Header="Edit" />
>     <MenuItem Header="Delete" />
>   </CompositeCollection>
> </UserControl.Resources>
> 
> <ContextMenu>
>   <ContextMenu.ItemsSource>
>     <CompositeCollection >
>       <CollectionContainer Collection="{StaticResource BaseMenu}" />
>       <Separator />
>       <MenuItem Header="Hello World" />
>       <MenuItem Header="Clickety Click">
>         <MenuItem Header="Clackety Clack" />
>       </MenuItem>
>     </CompositeCollection>
>   </ContextMenu.ItemsSource>
> </ContextMenu>
> ```

I came across this issue because of the situation that I do not want to define the same menu structure over and over again. Therefore I have created a `Base` menu

**Comment by MichelMichels** [comment]
Yes, I understand what you are trying to do. My last comment was explaining why this is happening. 

There might be some confusion about `MenuBase`:
* In your example, you are using `BaseMenu` as a name for a resource
* In my comment `MenuBase` is an abstract class that `Menu` and `ContextMenu` both inherit

To be clear, this explanation above is not the issue.

The issue is, as you mentioned,  that `CompositeCollection` is not a part of the Visual Tree. That is why following XAML line defined in the `MaterialDesignMenuItem` style is failing:

```xml
<Setter Property="Height" Value="{Binding Path=(wpf:MenuAssist.TopLevelMenuItemHeight), RelativeSource={RelativeSource Mode=FindAncestor, AncestorType=MenuBase}}" />
```

This line expects there to be a `UserControl` of class `MenuBase` to be in the Ancestors of the `MenuItem`; which is not the case due to the `CompositeCollection`.

**Comment by michaelmairegger** [comment]
> This line expects there to be a UserControl of .... due to the CompositeCollection.

Yes, unfortunately this is true


> There might be some confusion about MenuBase:
No, it was clear that my `BaseMenu` as ResourceKey has nothing to do with `x:Type MenuBase`.
I just wantet to state why I have such a situation
